CRDValidationRatcheting: Factor object correlation and comparison into reusable component

alexzielenski opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

CachedDeepEqual is needed for #121016 to use with the CEL validator. It will also be neeed for native types in declarative validation. This PR factors out the object correlation and comparison tree into a re-usable component. At validation time the instance can be shared by all validation passes.

Tried to keep all commits semantically a no-op

#### Which issue(s) this PR fixes:
N/A

#### Special notes for your reviewer:

/assign @apelisse 
/sig api-machinery

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://kep.k8s.io/4008
```

alexzielenski committed:
refactor: remove unnecessary lambda (noop)

alexzielenski committed:
refactor: directly implement ValueValidator

inlineValidator unnecessary, since we already have access to the `new` object

alexzielenski committed:
refactor: factor out object correlation

so it is reusable by CEL validators

alexzielenski committed:
refactor: export correlated fields

alexzielenski committed:
refactor: use common.Schema in CorrelatedObject

for reuseability with CEL structural schema impl

alexzielenski committed:
refactor: add methods for Key and Index on CorrelatedObject

going to be used from CEL validator too

alexzielenski committed:
refactor: move correlatedObject to its own file

no changes except package naming

jiahuif commented:
/triage accepted

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go:
```diff
@@ -408,3 +382,87 @@ func (f ratchetingValueValidator) SetPath(path string) {
+// Returns nil if the given name is does not exist in the new object, or its
+// value is not correlatable to an old value.
+// If receiver is nil or if the new value is not an object/map, returns nil.
+func (l *CorrelatedObject) Key(field string) *CorrelatedObject {
```
Do we need tests for these methods? (I'm looking commit-by-commit so sorry if that's done later).

alexzielenski replied:
tests added

apelisse commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go:
```diff
@@ -408,3 +382,87 @@ func (f ratchetingValueValidator) SetPath(path string) {
+		return nil
+	}
+
+	var propertySchema *spec.Schema
```
Will this work against non-structural schemas? Do you expect the schema to be structural here?

alexzielenski replied:
I'm not sure that matters since we are only traversing the structural nodes of the schema. In later commit this is refactored for `common.Schema` and does indeed work against structural

apelisse reviewed (commented):
Thanks! I don't see a lot of tests throughout, maybe they didn't need to be updated, if so that's great!

alexzielenski commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/ratcheting.go:
```diff
@@ -145,7 +145,7 @@ func (r *ratchetingValueValidator) getValidateOption() validate.Option {
 // This call has a side-effect of populating it's `children` variable with
 // the explored nodes of the object tree.
-func (r *ratchetingValueValidator) Validate() *validate.Result {
+func (r *ratchetingValueValidator) Validate(new interface{}) *validate.Result {
```
We don't care about new because we already know what it is as a consequence of doing correlation. It should be the same.

The reason it is required as a parameter is to conform to the `ValueValidator` interface so that it can be returned directly by `SubPropertyValidator` and `SubIndexValidator`. We also ignored `new` in the previous edition of the code so it is not a semantic change.

apelisse replied:
I'm not fan of this kind of things, any chance we can make it so that we don't have to use "new" when we create the validator but use the value given as a parameter instead?

alexzielenski replied:
It's not optimal, but I dont see a good way around it. I dont think would make sense for `CorrelatedObject.Key` and `Correlated.Index` to change to require the `new` when they don't need it, and this API works well for how I want to reuse the comparison for CEL and Schemas.

It is against the contract of `SubPropertyValidator` to call `Validate` on anything but the given property, so I think it is safe to do (same for index). It is `ratchetingValueValidator` is a completely internal implementation that is destroyed at the end of the call to `ValidateUpdate`

alexzielenski committed:
cleanup: add header and fix spelling

alexzielenski committed:
test: add correlatedobject test cases

alexzielenski committed:
refactor: make CachedDeepEqual independent of validation

before it required running validation first, now it builds the tree as needed

alexzielenski committed:
test: few more correlatedobject test cases

apelisse commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+// Returns nil if the given index is out of bounds, or its value is not
+// correlatable to an old value.
+// If receiver is nil or if the new value is not an array, returns nil.
+func (r *CorrelatedObject) Index(i int) *CorrelatedObject {
```
Should you use associative list keys rather than index to correlate items in a list?

alexzielenski replied:
This does use associative list keys when available.

apelisse commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality_test.go:
```diff
@@ -0,0 +1,702 @@
+                          type: string
+                        val:
+                          type: string
+                    x-kubernetes-list-type: atomic
```
All lists are atomic by default.

apelisse commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality_test.go:
```diff
@@ -0,0 +1,702 @@
+            `),
+			Schema: mustSchema(`
+                items:
+                    type: string
```
That list is supposed to be atomic.

alexzielenski replied:
why

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality_test.go:
```diff
@@ -0,0 +1,702 @@
+			return fmt.Errorf("expected old value %v, got %v", c.OldValue, correlatedObject.OldValue)
+		}
+
+		// Check that the correlated object is considered equal to the expected value
+		if (c.ExpectEqual || reflect.DeepEqual(correlatedObject.Value, correlatedObject.OldValue)) != correlatedObject.CachedDeepEqual() {
+			return fmt.Errorf("expected equal, got not equal")
+		}
```
No, the condition is correct. This is to account for cases where the object should be considered unchanged due to a re-ordering of an associative list.

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality_test.go:
```diff
@@ -0,0 +1,702 @@
+	// Create the correlated object
+	correlatedObject := common.NewCorrelatedObject(c.RootObject, c.RootOldObject, c.Schema)
+
+	// Traverse the correlated object
+	var err error
+	for _, key := range c.KeyPath {
+		if correlatedObject == nil {
+			break
+		}
+
+		switch k := key.(type) {
+		case string:
+			correlatedObject = correlatedObject.Key(k)
+		case int:
+			correlatedObject = correlatedObject.Index(k)
+		default:
+			return errors.New("key must be a string or int")
+		}
+		if err != nil {
+			return err
+		}
+	}
```
The purpose is to test:

1.  Following `Key/Index` yields expected correlated node
2. DeepEqual on the traversed node gives correct output

apelisse commented:
Given this is a refactor for alpha, let's move on. Would you mind fixing the broken test?

alexzielenski committed:
test: fix boilerplate

alexzielenski commented:
>  Given this is a refactor for alpha, let's move on. Would you mind fixing the broken test?

Fixed. We can address the rest of the testing nits after for test freeze, since time is so short to get implementation in.

alexzielenski commented:
/retest

apelisse commented:
/lgtm
/approve
/retest

alexzielenski commented:
/assign jpbetz

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+}
+
+// If oldValue is not a list, returns nil
+// If oldValue is a list takes mapType into account and attempts to find the
```
```suggestion
// If oldValue is a list , this function takes mapType into account and attempts to find the
```

alexzielenski replied:
Added

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+	}
+}
+
+// If oldValue is not a list, returns nil
```
```suggestion
// If OldValue or Value is not a list, or the index is out of bounds of the Value list  returns nil
```

alexzielenski replied:
Added

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+
+// If oldValue is not a list, returns nil
+// If oldValue is a list takes mapType into account and attempts to find the
+// old value with the same index or key, depending upon the mapType.
```
```suggestion
// old value with the same index or key, depending upon the mapType, and return it.
// If the old value is not found, returns nil.
```

alexzielenski replied:
Added

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+		}
+		return oldList.Get(currentElement)
+
+	case "set":
+		// Are sets correlatable? Only if the old value equals the current value.
+		// We might be able to support this, but do not currently see a lot
+		// of value
+		// (would allow you to add/remove items from sets with ratcheting but not change them)
+		return nil
```
Raise an error/panic or otherwise document that this is unsupported?

alexzielenski replied:
added documention that uncorrelatable lists like `set` return nil

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+		// Atomic lists are not correlatable by item
+		// Ratcheting is not available on a per-index basis
+		return nil
+	default:
```
label this as `case "granular"` since that is what it's for and leave a default with panic as the fall-through if somehow this function ever receives a list type we don't expect?

alexzielenski replied:
@apelisse pointed out the default list type was `atomic`, so this `default` case treating it as granular may be a bug.

This was also pointed out by @apelisse when the code originally went in, but we decided to punt whether we should treat list items as granular for the purposes of DeepEqual comparison in ratcheting. 

Thoughts @jpbetz? Should we change this to be consistent with SSA/CEL and treat arrays items unseparable? Leaving this lets users ratchet an array by modifying single elements, but not adding/removing any.

apelisse replied:
> only map-type is ratchetable. Same as CEL

That's the way to go. You can't correlate atomic, and it's either atomic or map-type/set. Mapping by index doesn't really make sense to me.

jpbetz replied:
I'm okay with deferring ratcheting on listType=atomic.  I see the main benefits being cases where we're working on values nested under a listType=map.

Good catch on the default=atomic BTW.  We should strive to always structure switch statements to reflect that.

alexzielenski replied:
added case for "", and made disallowed default case ratcheting

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+// old value with the same index or key, depending upon the mapType.
+//
+// If listType is map, creates a map representation of the list using the designated
+// map-keys and caches it for future calls.
```
```suggestion
// map-keys, caches it for future calls, and returns the map value, or nil if the correlated key is not in the old map.
```

alexzielenski replied:
added

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+	oldAsArray, oldIsArray := r.OldValue.([]interface{})
+	newAsArray, newIsArray := r.Value.([]interface{})
+
+	if oldIsArray != newIsArray {
+		return false
+	} else if oldIsArray {
```
Do we want to proceed if neither are lists? I looks like that if neither are lists this will return true? Is that what we want? I was expecting `if oldIsArray && newIsArray` here.

alexzielenski replied:
At this `else` condition we know that `oldIsArray == newIsArray`, so `oldIsArray` being true implies `newIsArray` is true and is equivalent to `oldIsArray && newIsArray`.

I've refactoerd this method to use a `switch` and commented some details like this

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+// value is not correlatable to an old value.
+// If receiver is nil or if the new value is not an object/map, returns nil.
+func (r *CorrelatedObject) Key(field string) *CorrelatedObject {
+	if r == nil || r.Schema == nil {
```
Should the other receiver functions also check if `r` is nil? I'm okay either way (all receivers check or no receivers check), I just noticed that this apparent inconsistent.  If this is done to support recursive descent, document?

alexzielenski replied:
Yeah it is done to support recursive descent. Added documentation and made all methods consistent

jpbetz reviewed (changes requested):
Nothing major, just some godoc suggestions and switch handling suggestions.

alexzielenski committed:
cleanup: add godoc

alexzielenski committed:
cleanup: consistent interface{} and any

alexzielenski committed:
cleanup: clarify correlatedOldValueForChildAtNewIndex comment

alexzielenski committed:
cleanup: consistently support nil receiver and document

alexzielenski committed:
cleanup: use swtich in CachedDeepEqual and add more comments

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+	"reflect"
+)
+
+type CorrelatedObject struct {
```
Added

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/cel/common/equality.go:
```diff
@@ -0,0 +1,275 @@
+//
+// If listType is map, creates a map representation of the list using the designated
+// map-keys and caches it for future calls.
+func (r *CorrelatedObject) correlateOldValueForChildAtNewIndex(index int) any {
```
changed to interface{} since it was the only occurrance of `any`

alexzielenski committed:
comments: clear up correlateOldValueForChildAtNewIndex godoc

alexzielenski committed:
ratcheting: disable correlation by index

discussion: https://github.com/kubernetes/kubernetes/pull/121118#discussion_r1358865893

alexzielenski commented:
Updated with latest feedback. @apelisse PTAL

apelisse reviewed (commented):
/lgtm

jpbetz commented:
/approve
